# Chuẩn bị dữ liệu

[Notebook gốc từ *Data Science: Introduction to Machine Learning for Data Science Python and Machine Learning Studio của Lee Stott*](https://github.com/leestott/intro-Datascience/blob/master/Course%20Materials/4-Cleaning_and_Manipulating-Reference.ipynb)

## Khám phá thông tin `DataFrame`

> **Mục tiêu học tập:** Sau phần này, bạn sẽ cảm thấy thoải mái khi tìm kiếm thông tin tổng quát về dữ liệu được lưu trữ trong pandas DataFrames.

Khi bạn đã tải dữ liệu vào pandas, rất có thể dữ liệu sẽ được lưu trong một `DataFrame`. Tuy nhiên, nếu tập dữ liệu trong `DataFrame` của bạn có 60.000 dòng và 400 cột, làm thế nào để bạn bắt đầu hiểu được dữ liệu mà mình đang làm việc? May mắn thay, pandas cung cấp một số công cụ tiện lợi để nhanh chóng xem thông tin tổng quát về một `DataFrame` ngoài việc xem vài dòng đầu tiên và cuối cùng.

Để khám phá chức năng này, chúng ta sẽ nhập thư viện scikit-learn của Python và sử dụng một tập dữ liệu mang tính biểu tượng mà mọi nhà khoa học dữ liệu đều đã thấy hàng trăm lần: tập dữ liệu *Iris* của nhà sinh học người Anh Ronald Fisher được sử dụng trong bài báo năm 1936 của ông "The use of multiple measurements in taxonomic problems":


In [1]:
import pandas as pd
from sklearn.datasets import load_iris

iris = load_iris()
iris_df = pd.DataFrame(data=iris['data'], columns=iris['feature_names'])

### `DataFrame.shape`
Chúng tôi đã tải bộ dữ liệu Iris vào biến `iris_df`. Trước khi đi sâu vào dữ liệu, sẽ rất hữu ích nếu biết số lượng điểm dữ liệu chúng tôi có và kích thước tổng thể của bộ dữ liệu. Việc xem xét khối lượng dữ liệu mà chúng tôi đang xử lý là rất cần thiết.


In [2]:
iris_df.shape

(150, 4)

Vậy, chúng ta đang xử lý 150 hàng và 4 cột dữ liệu. Mỗi hàng đại diện cho một điểm dữ liệu và mỗi cột đại diện cho một đặc điểm duy nhất liên quan đến khung dữ liệu. Nói cách khác, có 150 điểm dữ liệu, mỗi điểm chứa 4 đặc điểm.

`shape` ở đây là một thuộc tính của khung dữ liệu chứ không phải là một hàm, vì vậy nó không kết thúc bằng một cặp dấu ngoặc đơn.


### `DataFrame.columns`
Bây giờ chúng ta sẽ tìm hiểu về 4 cột dữ liệu. Mỗi cột thực sự đại diện cho điều gì? Thuộc tính `columns` sẽ cung cấp cho chúng ta tên của các cột trong dataframe.


In [3]:
iris_df.columns

Index(['sepal length (cm)', 'sepal width (cm)', 'petal length (cm)',
       'petal width (cm)'],
      dtype='object')

Như chúng ta có thể thấy, có bốn (4) cột. Thuộc tính `columns` cho chúng ta biết tên của các cột và về cơ bản không có gì khác. Thuộc tính này trở nên quan trọng khi chúng ta muốn xác định các đặc điểm mà một tập dữ liệu chứa.


### `DataFrame.info`
Số lượng dữ liệu (được cung cấp bởi thuộc tính `shape`) và tên của các đặc điểm hoặc cột (được cung cấp bởi thuộc tính `columns`) cho chúng ta một số thông tin về tập dữ liệu. Bây giờ, chúng ta sẽ muốn tìm hiểu sâu hơn về tập dữ liệu. Hàm `DataFrame.info()` rất hữu ích cho việc này.


In [4]:
iris_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150 entries, 0 to 149
Data columns (total 4 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   sepal length (cm)  150 non-null    float64
 1   sepal width (cm)   150 non-null    float64
 2   petal length (cm)  150 non-null    float64
 3   petal width (cm)   150 non-null    float64
dtypes: float64(4)
memory usage: 4.8 KB


Từ đây, chúng ta có thể rút ra một vài quan sát:  
1. Kiểu dữ liệu của mỗi cột: Trong tập dữ liệu này, tất cả dữ liệu đều được lưu dưới dạng số thực 64-bit.  
2. Số lượng giá trị không null: Xử lý các giá trị null là một bước quan trọng trong việc chuẩn bị dữ liệu. Việc này sẽ được thực hiện sau trong notebook.  


### DataFrame.describe()
Giả sử chúng ta có rất nhiều dữ liệu số trong tập dữ liệu của mình. Các phép tính thống kê đơn biến như giá trị trung bình, trung vị, tứ phân vị, v.v. có thể được thực hiện trên từng cột riêng lẻ. Hàm `DataFrame.describe()` cung cấp cho chúng ta một bản tóm tắt thống kê của các cột số trong tập dữ liệu.


In [5]:
iris_df.describe()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
count         150.000000        150.000000         150.000000        150.000000
mean            5.843333          3.057333           3.758000          1.199333
std             0.828066          0.435866           1.765298          0.762238
min             4.300000          2.000000           1.000000          0.100000
25%             5.100000          2.800000           1.600000          0.300000
50%             5.800000          3.000000           4.350000          1.300000
75%             6.400000          3.300000           5.100000          1.800000
max             7.900000          4.400000           6.900000          2.500000

Kết quả trên hiển thị tổng số điểm dữ liệu, giá trị trung bình, độ lệch chuẩn, giá trị nhỏ nhất, tứ phân vị dưới (25%), trung vị (50%), tứ phân vị trên (75%) và giá trị lớn nhất của mỗi cột.


### `DataFrame.head`
Với tất cả các hàm và thuộc tính đã đề cập ở trên, chúng ta đã có cái nhìn tổng quan về tập dữ liệu. Chúng ta biết có bao nhiêu điểm dữ liệu, có bao nhiêu đặc trưng, kiểu dữ liệu của mỗi đặc trưng và số lượng giá trị không null của mỗi đặc trưng.

Bây giờ là lúc xem xét chính dữ liệu. Hãy xem những hàng đầu tiên (những điểm dữ liệu đầu tiên) của `DataFrame` trông như thế nào:


In [6]:
iris_df.head()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
0                5.1               3.5                1.4               0.2
1                4.9               3.0                1.4               0.2
2                4.7               3.2                1.3               0.2
3                4.6               3.1                1.5               0.2
4                5.0               3.6                1.4               0.2

Như đầu ra ở đây, chúng ta có thể thấy năm (5) mục của tập dữ liệu. Nếu chúng ta nhìn vào chỉ số ở bên trái, chúng ta sẽ thấy rằng đây là năm hàng đầu tiên.


### Bài tập:

Từ ví dụ được đưa ra ở trên, rõ ràng rằng mặc định, `DataFrame.head` trả về năm hàng đầu tiên của một `DataFrame`. Trong ô mã dưới đây, bạn có thể tìm ra cách hiển thị nhiều hơn năm hàng không?


In [7]:
# Hint: Consult the documentation by using iris_df.head?

### `DataFrame.tail`
Một cách khác để xem dữ liệu là từ cuối (thay vì từ đầu). Đối ngược với `DataFrame.head` là `DataFrame.tail`, nó trả về năm hàng cuối cùng của một `DataFrame`:


In [8]:
iris_df.tail()

sepal length (cm)  sepal width (cm)  petal length (cm)  petal width (cm)
145                6.7               3.0                5.2               2.3
146                6.3               2.5                5.0               1.9
147                6.5               3.0                5.2               2.0
148                6.2               3.4                5.4               2.3
149                5.9               3.0                5.1               1.8

Trong thực tế, việc dễ dàng kiểm tra vài hàng đầu tiên hoặc vài hàng cuối cùng của một `DataFrame` rất hữu ích, đặc biệt khi bạn đang tìm kiếm các giá trị ngoại lệ trong các tập dữ liệu có thứ tự.

Tất cả các hàm và thuộc tính được trình bày ở trên với sự hỗ trợ của các ví dụ mã đều giúp chúng ta có cái nhìn tổng quan về dữ liệu.

> **Điểm rút ra:** Chỉ cần nhìn vào siêu dữ liệu về thông tin trong một DataFrame hoặc vài giá trị đầu tiên và cuối cùng của nó, bạn có thể ngay lập tức hình dung về kích thước, hình dạng và nội dung của dữ liệu mà bạn đang xử lý.


### Dữ liệu bị thiếu
Hãy cùng tìm hiểu về dữ liệu bị thiếu. Dữ liệu bị thiếu xảy ra khi không có giá trị nào được lưu trữ trong một số cột.

Hãy lấy một ví dụ: giả sử ai đó rất quan tâm đến cân nặng của mình và không điền vào trường cân nặng trong một cuộc khảo sát. Khi đó, giá trị cân nặng của người đó sẽ bị thiếu.

Phần lớn thời gian, trong các tập dữ liệu thực tế, giá trị bị thiếu thường xảy ra.

**Cách Pandas xử lý dữ liệu bị thiếu**

Pandas xử lý giá trị bị thiếu theo hai cách. Cách đầu tiên bạn đã thấy trong các phần trước: `NaN`, hoặc Not a Number. Đây thực chất là một giá trị đặc biệt thuộc về đặc tả số học dấu phẩy động IEEE và chỉ được sử dụng để biểu thị các giá trị dấu phẩy động bị thiếu.

Đối với các giá trị bị thiếu không phải là số dấu phẩy động, pandas sử dụng đối tượng `None` của Python. Mặc dù có thể gây nhầm lẫn khi bạn gặp hai loại giá trị khác nhau nhưng đều biểu thị cùng một ý nghĩa, có những lý do lập trình hợp lý cho sự lựa chọn thiết kế này. Trên thực tế, cách tiếp cận này cho phép pandas cung cấp một sự thỏa hiệp tốt cho phần lớn các trường hợp. Tuy nhiên, cả `None` và `NaN` đều có những hạn chế mà bạn cần lưu ý liên quan đến cách chúng có thể được sử dụng.


### `None`: dữ liệu thiếu không phải kiểu float
Vì `None` xuất phát từ Python, nên nó không thể được sử dụng trong các mảng NumPy và pandas không có kiểu dữ liệu `'object'`. Hãy nhớ rằng, các mảng NumPy (và các cấu trúc dữ liệu trong pandas) chỉ có thể chứa một loại dữ liệu duy nhất. Đây chính là điều mang lại sức mạnh vượt trội cho chúng trong việc xử lý dữ liệu quy mô lớn và tính toán, nhưng đồng thời cũng hạn chế sự linh hoạt của chúng. Những mảng như vậy phải được chuyển đổi lên kiểu dữ liệu “mẫu số chung thấp nhất,” tức là kiểu dữ liệu có thể bao gồm tất cả các phần tử trong mảng. Khi `None` xuất hiện trong mảng, điều đó có nghĩa là bạn đang làm việc với các đối tượng Python.

Để thấy điều này trong thực tế, hãy xem xét mảng ví dụ sau (lưu ý `dtype` của nó):


In [9]:
import numpy as np

example1 = np.array([2, None, 6, 8])
example1

array([2, None, 6, 8], dtype=object)

Thực tế của việc nâng cấp kiểu dữ liệu mang theo hai tác động phụ. Thứ nhất, các thao tác sẽ được thực hiện ở mức mã Python được diễn giải thay vì mã NumPy đã biên dịch. Về cơ bản, điều này có nghĩa là bất kỳ thao tác nào liên quan đến `Series` hoặc `DataFrame` có chứa `None` sẽ chậm hơn. Mặc dù bạn có thể không nhận ra sự giảm hiệu suất này, nhưng với các tập dữ liệu lớn, điều này có thể trở thành một vấn đề.

Tác động phụ thứ hai bắt nguồn từ tác động phụ đầu tiên. Vì `None` về cơ bản kéo `Series` hoặc `DataFrame` trở lại thế giới của Python thuần túy, việc sử dụng các phép tổng hợp của NumPy/pandas như `sum()` hoặc `min()` trên các mảng chứa giá trị ``None`` thường sẽ gây ra lỗi:


In [10]:
example1.sum()

TypeError: ignored

**Điểm chính**: Phép cộng (và các phép toán khác) giữa các số nguyên và giá trị `None` là không xác định, điều này có thể hạn chế những gì bạn có thể làm với các tập dữ liệu chứa chúng.


### `NaN`: giá trị float bị thiếu

Khác với `None`, NumPy (và do đó là pandas) hỗ trợ `NaN` cho các thao tác vector hóa nhanh và ufuncs. Tin xấu là bất kỳ phép toán nào thực hiện trên `NaN` luôn cho kết quả là `NaN`. Ví dụ:


In [11]:
np.nan + 1

nan

In [12]:
np.nan * 0

nan

Tin tốt: các phép tổng hợp chạy trên mảng có chứa `NaN` không gây lỗi. Tin xấu: kết quả không phải lúc nào cũng hữu ích:


In [13]:
example2 = np.array([2, np.nan, 6, 8]) 
example2.sum(), example2.min(), example2.max()

(nan, nan, nan)

### Bài tập:


In [11]:
# What happens if you add np.nan and None together?


Hãy nhớ: `NaN` chỉ dành cho các giá trị dấu phẩy động bị thiếu; không có tương đương `NaN` cho số nguyên, chuỗi hoặc Boolean.


### `NaN` và `None`: giá trị null trong pandas

Mặc dù `NaN` và `None` có thể hoạt động hơi khác nhau, pandas vẫn được thiết kế để xử lý chúng thay thế cho nhau. Để hiểu rõ hơn, hãy xem xét một `Series` gồm các số nguyên:


In [15]:
int_series = pd.Series([1, 2, 3], dtype=int)
int_series

0    1
1    2
2    3
dtype: int64

### Bài tập:


In [16]:
# Now set an element of int_series equal to None.
# How does that element show up in the Series?
# What is the dtype of the Series?


Trong quá trình nâng cấp kiểu dữ liệu để thiết lập tính đồng nhất dữ liệu trong `Series` và `DataFrame`, pandas sẽ sẵn sàng chuyển đổi giá trị thiếu giữa `None` và `NaN`. Do đặc điểm thiết kế này, sẽ hữu ích nếu bạn nghĩ về `None` và `NaN` như hai dạng khác nhau của "null" trong pandas. Thực tế, một số phương thức cốt lõi mà bạn sẽ sử dụng để xử lý giá trị thiếu trong pandas phản ánh ý tưởng này qua tên gọi của chúng:

- `isnull()`: Tạo một mặt nạ Boolean chỉ ra các giá trị thiếu
- `notnull()`: Ngược lại với `isnull()`
- `dropna()`: Trả về một phiên bản dữ liệu đã được lọc
- `fillna()`: Trả về một bản sao của dữ liệu với các giá trị thiếu được điền hoặc ước lượng

Đây là những phương thức quan trọng cần nắm vững và làm quen, vì vậy hãy cùng tìm hiểu chi tiết từng phương thức.


### Phát hiện giá trị null

Sau khi đã hiểu được tầm quan trọng của các giá trị thiếu, chúng ta cần phát hiện chúng trong tập dữ liệu trước khi xử lý. 
Cả `isnull()` và `notnull()` đều là các phương pháp chính để phát hiện dữ liệu null. Cả hai đều trả về mặt nạ Boolean trên dữ liệu của bạn.


In [17]:
example3 = pd.Series([0, np.nan, '', None])

In [18]:
example3.isnull()

0    False
1     True
2    False
3     True
dtype: bool

Hãy nhìn kỹ vào kết quả. Có điều gì khiến bạn ngạc nhiên không? Mặc dù `0` là một giá trị null trong toán học, nhưng nó vẫn là một số nguyên hoàn toàn hợp lệ và pandas xử lý nó như vậy. `''` thì tinh tế hơn một chút. Mặc dù chúng ta đã sử dụng nó trong Phần 1 để biểu thị một giá trị chuỗi rỗng, nhưng nó vẫn là một đối tượng chuỗi và không được pandas coi là một biểu diễn của null.

Bây giờ, hãy đảo ngược lại và sử dụng các phương pháp này theo cách giống như cách bạn sẽ sử dụng chúng trong thực tế. Bạn có thể sử dụng các mặt nạ Boolean trực tiếp như một chỉ mục của ``Series`` hoặc ``DataFrame``, điều này có thể hữu ích khi cố gắng làm việc với các giá trị bị thiếu (hoặc hiện diện) riêng lẻ.

Nếu chúng ta muốn tổng số giá trị bị thiếu, chúng ta chỉ cần thực hiện phép cộng trên mặt nạ được tạo bởi phương pháp `isnull()`.


In [19]:
example3.isnull().sum()

2

### Bài tập:


In [20]:
# Try running example3[example3.notnull()].
# Before you do so, what do you expect to see?


**Điểm chính**: Cả hai phương thức `isnull()` và `notnull()` đều tạo ra kết quả tương tự khi bạn sử dụng chúng trong DataFrames: chúng hiển thị kết quả và chỉ mục của những kết quả đó, điều này sẽ giúp bạn rất nhiều khi xử lý dữ liệu của mình.


### Xử lý dữ liệu thiếu

> **Mục tiêu học tập:** Sau khi hoàn thành phần này, bạn sẽ biết cách và thời điểm thay thế hoặc loại bỏ các giá trị null trong DataFrames.

Các mô hình Machine Learning không thể tự xử lý dữ liệu thiếu. Vì vậy, trước khi đưa dữ liệu vào mô hình, chúng ta cần xử lý các giá trị thiếu này.

Cách xử lý dữ liệu thiếu mang lại những sự đánh đổi tinh tế, có thể ảnh hưởng đến phân tích cuối cùng và kết quả thực tế.

Có hai cách chính để xử lý dữ liệu thiếu:

1.   Loại bỏ hàng chứa giá trị thiếu
2.   Thay thế giá trị thiếu bằng một giá trị khác

Chúng ta sẽ thảo luận chi tiết về cả hai phương pháp này cùng với ưu và nhược điểm của chúng.


### Loại bỏ giá trị null

Lượng dữ liệu chúng ta đưa vào mô hình có ảnh hưởng trực tiếp đến hiệu suất của nó. Việc loại bỏ giá trị null đồng nghĩa với việc giảm số lượng điểm dữ liệu, từ đó giảm kích thước của tập dữ liệu. Vì vậy, nên loại bỏ các hàng có giá trị null khi tập dữ liệu khá lớn.

Một trường hợp khác có thể xảy ra là một hàng hoặc cột nào đó có rất nhiều giá trị bị thiếu. Khi đó, chúng có thể bị loại bỏ vì chúng không mang lại nhiều giá trị cho phân tích của chúng ta do phần lớn dữ liệu bị thiếu ở hàng/cột đó.

Ngoài việc xác định các giá trị bị thiếu, pandas cung cấp một cách tiện lợi để loại bỏ giá trị null từ `Series` và `DataFrame`. Để thấy điều này trong thực tế, hãy quay lại `example3`. Hàm `DataFrame.dropna()` giúp loại bỏ các hàng có giá trị null.


In [21]:
example3 = example3.dropna()
example3

0    0
2     
dtype: object

Lưu ý rằng điều này nên giống như kết quả từ `example3[example3.notnull()]`. Sự khác biệt ở đây là, thay vì chỉ lập chỉ mục trên các giá trị bị che, `dropna` đã loại bỏ những giá trị bị thiếu khỏi `Series` `example3`.

Vì DataFrames có hai chiều, chúng cung cấp nhiều tùy chọn hơn để loại bỏ dữ liệu.


In [22]:
example4 = pd.DataFrame([[1,      np.nan, 7], 
                         [2,      5,      8], 
                         [np.nan, 6,      9]])
example4

0    1  2
0  1.0  NaN  7
1  2.0  5.0  8
2  NaN  6.0  9

(Bạn có nhận thấy rằng pandas đã chuyển đổi hai cột thành kiểu float để phù hợp với các giá trị `NaN` không?)

Bạn không thể xóa một giá trị đơn lẻ từ một `DataFrame`, vì vậy bạn phải xóa toàn bộ hàng hoặc cột. Tùy thuộc vào những gì bạn đang làm, bạn có thể muốn thực hiện một trong hai cách, và pandas cung cấp cho bạn tùy chọn cho cả hai. Vì trong khoa học dữ liệu, các cột thường đại diện cho biến số và các hàng đại diện cho quan sát, bạn có khả năng sẽ xóa các hàng dữ liệu hơn; cài đặt mặc định của `dropna()` là xóa tất cả các hàng chứa bất kỳ giá trị null nào:


In [23]:
example4.dropna()

0    1  2
1  2.0  5.0  8

Nếu cần thiết, bạn có thể loại bỏ các giá trị NA từ các cột. Sử dụng `axis=1` để thực hiện điều này:


In [24]:
example4.dropna(axis='columns')

2
0  7
1  8
2  9

Lưu ý rằng điều này có thể loại bỏ rất nhiều dữ liệu mà bạn có thể muốn giữ lại, đặc biệt là trong các tập dữ liệu nhỏ. Vậy nếu bạn chỉ muốn loại bỏ các hàng hoặc cột chứa một số hoặc thậm chí tất cả các giá trị null thì sao? Bạn có thể chỉ định các thiết lập đó trong `dropna` với các tham số `how` và `thresh`.

Theo mặc định, `how='any'` (nếu bạn muốn tự kiểm tra hoặc xem các tham số khác mà phương thức này có, hãy chạy `example4.dropna?` trong một ô mã). Ngoài ra, bạn có thể chỉ định `how='all'` để chỉ loại bỏ các hàng hoặc cột chứa tất cả các giá trị null. Hãy mở rộng ví dụ `DataFrame` của chúng ta để xem điều này hoạt động như thế nào trong bài tập tiếp theo.


In [25]:
example4[3] = np.nan
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

> Những điểm chính:  
1. Chỉ nên loại bỏ các giá trị null nếu tập dữ liệu đủ lớn.  
2. Có thể loại bỏ toàn bộ hàng hoặc cột nếu phần lớn dữ liệu của chúng bị thiếu.  
3. Phương thức `DataFrame.dropna(axis=)` hỗ trợ loại bỏ các giá trị null. Tham số `axis` xác định liệu sẽ loại bỏ hàng hay cột.  
4. Tham số `how` cũng có thể được sử dụng. Mặc định, nó được đặt là `any`. Vì vậy, nó chỉ loại bỏ những hàng/cột chứa bất kỳ giá trị null nào. Có thể đặt thành `all` để chỉ định rằng chúng ta sẽ chỉ loại bỏ những hàng/cột mà tất cả các giá trị đều là null.  


### Bài tập:


In [22]:
# How might you go about dropping just column 3?
# Hint: remember that you will need to supply both the axis parameter and the how parameter.


Tham số `thresh` cung cấp cho bạn khả năng kiểm soát chi tiết hơn: bạn đặt số lượng giá trị *không null* mà một hàng hoặc cột cần có để được giữ lại:


In [27]:
example4.dropna(axis='rows', thresh=3)

0    1  2   3
1  2.0  5.0  8 NaN

Ở đây, hàng đầu tiên và hàng cuối cùng đã bị loại bỏ, vì chúng chỉ chứa hai giá trị không null.


### Điền giá trị null

Đôi khi việc điền các giá trị thiếu bằng những giá trị có thể hợp lệ là điều hợp lý. Có một vài kỹ thuật để điền giá trị null. Kỹ thuật đầu tiên là sử dụng Kiến Thức Chuyên Ngành (kiến thức về chủ đề mà tập dữ liệu dựa trên) để ước lượng giá trị thiếu một cách nào đó.

Bạn có thể sử dụng `isnull` để thực hiện việc này trực tiếp, nhưng điều đó có thể rất tốn công, đặc biệt nếu bạn có nhiều giá trị cần điền. Vì đây là một nhiệm vụ phổ biến trong khoa học dữ liệu, pandas cung cấp `fillna`, trả về một bản sao của `Series` hoặc `DataFrame` với các giá trị thiếu được thay thế bằng giá trị bạn chọn. Hãy tạo một ví dụ khác về `Series` để xem cách hoạt động này trong thực tế.


### Dữ liệu phân loại (Không phải số)
Đầu tiên, hãy xem xét dữ liệu không phải số. Trong các tập dữ liệu, chúng ta có các cột chứa dữ liệu phân loại. Ví dụ: Giới tính, Đúng hoặc Sai, v.v.

Trong hầu hết các trường hợp này, chúng ta thay thế các giá trị bị thiếu bằng `mode` của cột. Giả sử, chúng ta có 100 điểm dữ liệu, trong đó 90 điểm là Đúng, 8 điểm là Sai và 2 điểm không được điền. Khi đó, chúng ta có thể điền 2 điểm bị thiếu bằng Đúng, dựa trên toàn bộ cột.

Một lần nữa, ở đây chúng ta có thể sử dụng kiến thức chuyên môn. Hãy xem xét một ví dụ về việc điền bằng mode.


In [28]:
fill_with_mode = pd.DataFrame([[1,2,"True"],
                               [3,4,None],
                               [5,6,"False"],
                               [7,8,"True"],
                               [9,10,"True"]])

fill_with_mode

0   1      2
0  1   2   True
1  3   4   None
2  5   6  False
3  7   8   True
4  9  10   True

Bây giờ, hãy tìm giá trị mode trước khi điền giá trị `None` bằng mode.


In [29]:
fill_with_mode[2].value_counts()

True     3
False    1
Name: 2, dtype: int64

Vì vậy, chúng ta sẽ thay thế None bằng True


In [30]:
fill_with_mode[2].fillna('True',inplace=True)

In [31]:
fill_with_mode

0   1      2
0  1   2   True
1  3   4   True
2  5   6  False
3  7   8   True
4  9  10   True

Như chúng ta có thể thấy, giá trị null đã được thay thế. Không cần phải nói, chúng ta có thể viết bất kỳ điều gì thay thế cho `'True'` và nó sẽ được thay thế.


### Dữ liệu số
Bây giờ, chúng ta sẽ nói về dữ liệu số. Ở đây, có hai cách phổ biến để thay thế giá trị bị thiếu:

1. Thay thế bằng giá trị trung vị của hàng
2. Thay thế bằng giá trị trung bình của hàng

Chúng ta sử dụng giá trị trung vị trong trường hợp dữ liệu bị lệch và có các giá trị ngoại lai. Điều này là do trung vị không bị ảnh hưởng bởi các giá trị ngoại lai.

Khi dữ liệu đã được chuẩn hóa, chúng ta có thể sử dụng giá trị trung bình, vì trong trường hợp này, giá trị trung bình và trung vị sẽ khá gần nhau.

Đầu tiên, hãy lấy một cột có phân phối chuẩn và điền giá trị bị thiếu bằng giá trị trung bình của cột.


In [32]:
fill_with_mean = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [np.nan,4,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  NaN  4  5
3  1.0  6  7
4  2.0  8  9

Giá trị trung bình của cột là


In [33]:
np.mean(fill_with_mean[0])

0.0

Điền bằng giá trị trung bình


In [34]:
fill_with_mean[0].fillna(np.mean(fill_with_mean[0]),inplace=True)
fill_with_mean

0  1  2
0 -2.0  0  1
1 -1.0  2  3
2  0.0  4  5
3  1.0  6  7
4  2.0  8  9

Như chúng ta có thể thấy, giá trị bị thiếu đã được thay thế bằng giá trị trung bình của nó.


Bây giờ hãy thử một dataframe khác, và lần này chúng ta sẽ thay thế các giá trị None bằng giá trị trung vị của cột.


In [35]:
fill_with_median = pd.DataFrame([[-2,0,1],
                               [-1,2,3],
                               [0,np.nan,5],
                               [1,6,7],
                               [2,8,9]])

fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  NaN  5
3  1  6.0  7
4  2  8.0  9

Trung vị của cột thứ hai là


In [36]:
fill_with_median[1].median()

4.0

Điền bằng giá trị trung vị


In [37]:
fill_with_median[1].fillna(fill_with_median[1].median(),inplace=True)
fill_with_median

0    1  2
0 -2  0.0  1
1 -1  2.0  3
2  0  4.0  5
3  1  6.0  7
4  2  8.0  9

Như chúng ta có thể thấy, giá trị NaN đã được thay thế bằng giá trị trung vị của cột.


In [38]:
example5 = pd.Series([1, np.nan, 2, None, 3], index=list('abcde'))
example5

a    1.0
b    NaN
c    2.0
d    NaN
e    3.0
dtype: float64

Bạn có thể điền tất cả các mục null bằng một giá trị duy nhất, chẳng hạn như `0`:


In [39]:
example5.fillna(0)

a    1.0
b    0.0
c    2.0
d    0.0
e    3.0
dtype: float64

> Những điểm chính cần ghi nhớ:
1. Việc điền giá trị thiếu nên được thực hiện khi dữ liệu còn ít hoặc khi có chiến lược để điền giá trị thiếu.
2. Kiến thức chuyên môn có thể được sử dụng để ước lượng và điền giá trị thiếu.
3. Đối với dữ liệu phân loại, thường giá trị thiếu được thay thế bằng giá trị mode của cột.
4. Đối với dữ liệu số, giá trị thiếu thường được điền bằng giá trị trung bình (đối với các tập dữ liệu đã chuẩn hóa) hoặc giá trị trung vị của cột.


### Bài tập:


In [40]:
# What happens if you try to fill null values with a string, like ''?


Bạn có thể **điền trước** các giá trị null, tức là sử dụng giá trị hợp lệ cuối cùng để điền vào null:


In [41]:
example5.fillna(method='ffill')

a    1.0
b    1.0
c    2.0
d    2.0
e    3.0
dtype: float64

Bạn cũng có thể **lấp đầy ngược** để truyền giá trị hợp lệ tiếp theo ngược lại nhằm lấp đầy một giá trị null:


In [42]:
example5.fillna(method='bfill')

a    1.0
b    2.0
c    2.0
d    3.0
e    3.0
dtype: float64

Như bạn có thể đoán, điều này hoạt động tương tự với DataFrames, nhưng bạn cũng có thể chỉ định một `axis` để điền các giá trị null:


In [43]:
example4

0    1  2   3
0  1.0  NaN  7 NaN
1  2.0  5.0  8 NaN
2  NaN  6.0  9 NaN

In [44]:
example4.fillna(method='ffill', axis=1)

0    1    2    3
0  1.0  1.0  7.0  7.0
1  2.0  5.0  8.0  8.0
2  NaN  6.0  9.0  9.0

Lưu ý rằng khi không có giá trị trước đó để điền tiếp, giá trị null vẫn được giữ nguyên.


### Bài tập:


In [45]:
# What output does example4.fillna(method='bfill', axis=1) produce?
# What about example4.fillna(method='ffill') or example4.fillna(method='bfill')?
# Can you think of a longer code snippet to write that can fill all of the null values in example4?


Bạn có thể sáng tạo về cách sử dụng `fillna`. Ví dụ, hãy xem lại `example4`, nhưng lần này hãy điền các giá trị thiếu bằng giá trị trung bình của tất cả các giá trị trong `DataFrame`:


In [46]:
example4.fillna(example4.mean())

0    1  2   3
0  1.0  5.5  7 NaN
1  2.0  5.0  8 NaN
2  1.5  6.0  9 NaN

Lưu ý rằng cột 3 vẫn chưa có giá trị: hướng mặc định là điền giá trị theo hàng.

> **Điểm cần nhớ:** Có nhiều cách để xử lý giá trị thiếu trong tập dữ liệu của bạn. Chiến lược cụ thể mà bạn sử dụng (loại bỏ, thay thế, hoặc thậm chí cách bạn thay thế) nên được quyết định dựa trên đặc điểm cụ thể của dữ liệu đó. Bạn sẽ phát triển cảm giác tốt hơn về cách xử lý giá trị thiếu khi bạn làm việc và tương tác nhiều hơn với các tập dữ liệu.


### Mã hóa Dữ liệu Phân loại

Các mô hình học máy chỉ xử lý dữ liệu dạng số và bất kỳ dạng dữ liệu số nào. Nó sẽ không thể phân biệt giữa "Có" và "Không", nhưng sẽ có thể phân biệt giữa 0 và 1. Vì vậy, sau khi điền các giá trị bị thiếu, chúng ta cần mã hóa dữ liệu phân loại thành một dạng số để mô hình có thể hiểu được.

Việc mã hóa có thể thực hiện theo hai cách. Chúng ta sẽ thảo luận về chúng ngay sau đây.


**MÃ HÓA NHÃN**

Mã hóa nhãn về cơ bản là chuyển đổi mỗi danh mục thành một số. Ví dụ, giả sử chúng ta có một tập dữ liệu về hành khách hàng không và có một cột chứa hạng ghế của họ trong số các hạng ['business class', 'economy class', 'first class']. Nếu thực hiện mã hóa nhãn, cột này sẽ được chuyển đổi thành [0,1,2]. Hãy cùng xem một ví dụ qua đoạn mã. Vì chúng ta sẽ học `scikit-learn` trong các notebook sắp tới, nên ở đây chúng ta sẽ không sử dụng nó.


In [47]:
label = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
label

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Để thực hiện mã hóa nhãn trên cột thứ nhất, chúng ta phải đầu tiên mô tả một ánh xạ từ mỗi lớp sang một số, trước khi thay thế.


In [48]:
class_labels = {'business class':0,'economy class':1,'first class':2}
label['class'] = label['class'].replace(class_labels)
label

ID  class
0  10      0
1  20      2
2  30      1
3  40      1
4  50      1
5  60      0

Như chúng ta có thể thấy, kết quả đầu ra khớp với những gì chúng ta dự đoán sẽ xảy ra. Vậy, khi nào chúng ta sử dụng mã hóa nhãn? Mã hóa nhãn được sử dụng trong một hoặc cả hai trường hợp sau:
1. Khi số lượng danh mục lớn
2. Khi các danh mục có thứ tự.


**MÃ HÓA ONE HOT**

Một loại mã hóa khác là One Hot Encoding. Trong loại mã hóa này, mỗi danh mục của cột sẽ được thêm vào dưới dạng một cột riêng biệt và mỗi điểm dữ liệu sẽ nhận giá trị 0 hoặc 1 dựa trên việc nó có chứa danh mục đó hay không. Vì vậy, nếu có n danh mục khác nhau, n cột sẽ được thêm vào dataframe.

Ví dụ, hãy lấy ví dụ về các hạng ghế máy bay. Các danh mục là: ['business class', 'economy class', 'first class']. Nếu chúng ta thực hiện mã hóa one hot, ba cột sau sẽ được thêm vào tập dữ liệu: ['class_business class', 'class_economy class', 'class_first class'].


In [49]:
one_hot = pd.DataFrame([
                      [10,'business class'],
                      [20,'first class'],
                      [30, 'economy class'],
                      [40, 'economy class'],
                      [50, 'economy class'],
                      [60, 'business class']
],columns=['ID','class'])
one_hot

ID           class
0  10  business class
1  20     first class
2  30   economy class
3  40   economy class
4  50   economy class
5  60  business class

Hãy thực hiện mã hóa one hot trên cột đầu tiên


In [50]:
one_hot_data = pd.get_dummies(one_hot,columns=['class'])

In [51]:
one_hot_data

ID  class_business class  class_economy class  class_first class
0  10                     1                    0                  0
1  20                     0                    0                  1
2  30                     0                    1                  0
3  40                     0                    1                  0
4  50                     0                    1                  0
5  60                     1                    0                  0

Mỗi cột được mã hóa one-hot chứa 0 hoặc 1, xác định liệu danh mục đó có tồn tại cho điểm dữ liệu đó hay không.


Khi nào chúng ta sử dụng mã hóa one hot? Mã hóa one hot được sử dụng trong một hoặc cả hai trường hợp sau:

1. Khi số lượng danh mục và kích thước của tập dữ liệu nhỏ.
2. Khi các danh mục không tuân theo thứ tự cụ thể nào.


> Những điểm chính:
1. Mã hóa được thực hiện để chuyển đổi dữ liệu không phải số thành dữ liệu số.  
2. Có hai loại mã hóa: Mã hóa nhãn và Mã hóa One Hot, cả hai đều có thể được thực hiện dựa trên yêu cầu của tập dữ liệu.  


## Loại bỏ dữ liệu trùng lặp

> **Mục tiêu học tập:** Sau khi hoàn thành phần này, bạn sẽ tự tin trong việc xác định và loại bỏ các giá trị trùng lặp từ DataFrames.

Ngoài dữ liệu bị thiếu, bạn thường gặp phải dữ liệu trùng lặp trong các tập dữ liệu thực tế. May mắn thay, pandas cung cấp một cách dễ dàng để phát hiện và loại bỏ các mục trùng lặp.


### Xác định giá trị trùng lặp: `duplicated`

Bạn có thể dễ dàng phát hiện các giá trị trùng lặp bằng phương pháp `duplicated` trong pandas, phương pháp này trả về một mặt nạ Boolean cho biết liệu một mục trong `DataFrame` có phải là bản sao của một mục trước đó hay không. Hãy tạo một ví dụ `DataFrame` khác để xem cách hoạt động của nó.


In [52]:
example6 = pd.DataFrame({'letters': ['A','B'] * 2 + ['B'],
                         'numbers': [1, 2, 1, 3, 3]})
example6

letters  numbers
0       A        1
1       B        2
2       A        1
3       B        3
4       B        3

In [53]:
example6.duplicated()

0    False
1    False
2     True
3    False
4     True
dtype: bool

### Loại bỏ các giá trị trùng lặp: `drop_duplicates`
`drop_duplicates` đơn giản trả về một bản sao của dữ liệu mà tất cả các giá trị `duplicated` đều là `False`:


In [54]:
example6.drop_duplicates()

letters  numbers
0       A        1
1       B        2
3       B        3

Cả `duplicated` và `drop_duplicates` mặc định xem xét tất cả các cột nhưng bạn có thể chỉ định rằng chúng chỉ kiểm tra một tập hợp con các cột trong `DataFrame` của bạn:


In [55]:
example6.drop_duplicates(['letters'])

letters  numbers
0       A        1
1       B        2

> **Điểm chính:** Loại bỏ dữ liệu trùng lặp là một phần thiết yếu của hầu hết mọi dự án khoa học dữ liệu. Dữ liệu trùng lặp có thể làm thay đổi kết quả phân tích của bạn và đưa ra kết quả không chính xác!


## Kiểm tra chất lượng dữ liệu thực tế

> **Mục tiêu học tập:** Sau phần này, bạn sẽ tự tin phát hiện và sửa chữa các vấn đề chất lượng dữ liệu phổ biến trong thực tế, bao gồm giá trị phân loại không nhất quán, giá trị số bất thường (ngoại lệ), và các thực thể trùng lặp với biến thể.

Mặc dù giá trị thiếu và các bản sao chính xác là những vấn đề phổ biến, các tập dữ liệu thực tế thường chứa những vấn đề tinh vi hơn:

1. **Giá trị phân loại không nhất quán**: Cùng một danh mục được viết khác nhau (ví dụ: "USA", "U.S.A", "United States")
2. **Giá trị số bất thường**: Các ngoại lệ cực đoan cho thấy lỗi nhập liệu (ví dụ: tuổi = 999)
3. **Các hàng gần giống nhau**: Các bản ghi đại diện cho cùng một thực thể nhưng có sự khác biệt nhỏ

Hãy cùng khám phá các kỹ thuật để phát hiện và xử lý những vấn đề này.


### Tạo một Bộ Dữ Liệu "Bẩn" Mẫu

Đầu tiên, hãy tạo một bộ dữ liệu mẫu chứa các loại vấn đề mà chúng ta thường gặp trong dữ liệu thực tế:


In [ ]:
import pandas as pd
import numpy as np

# Create a sample dataset with quality issues
dirty_data = pd.DataFrame({
    'customer_id': [1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12],
    'name': ['John Smith', 'Jane Doe', 'John Smith', 'Bob Johnson', 
             'Alice Williams', 'Charlie Brown', 'John  Smith', 'Eva Martinez',
             'Bob Johnson', 'Diana Prince', 'Frank Castle', 'Alice Williams'],
    'age': [25, 32, 25, 45, 28, 199, 25, 31, 45, 27, -5, 28],
    'country': ['USA', 'UK', 'U.S.A', 'Canada', 'USA', 'United Kingdom',
                'United States', 'Mexico', 'canada', 'USA', 'UK', 'usa'],
    'purchase_amount': [100.50, 250.00, 105.00, 320.00, 180.00, 90.00,
                       102.00, 275.00, 325.00, 195.00, 410.00, 185.00]
})

print("Sample 'Dirty' Dataset:")
print(dirty_data)

### 1. Phát hiện các giá trị danh mục không nhất quán

Lưu ý rằng cột `country` có nhiều cách biểu diễn cho cùng một quốc gia. Hãy xác định những điểm không nhất quán này:


In [ ]:
# Check unique values in the country column
print("Unique country values:")
print(dirty_data['country'].unique())
print(f"\nTotal unique values: {dirty_data['country'].nunique()}")

# Count occurrences of each variation
print("\nValue counts:")
print(dirty_data['country'].value_counts())

#### Chuẩn hóa các giá trị phân loại

Chúng ta có thể tạo một ánh xạ để chuẩn hóa các giá trị này. Một cách đơn giản là chuyển đổi sang chữ thường và tạo một từ điển ánh xạ:


In [ ]:
# Create a standardization mapping
country_mapping = {
    'usa': 'USA',
    'u.s.a': 'USA',
    'united states': 'USA',
    'uk': 'UK',
    'united kingdom': 'UK',
    'canada': 'Canada',
    'mexico': 'Mexico'
}

# Standardize the country column
dirty_data['country_clean'] = dirty_data['country'].str.lower().map(country_mapping)

print("Before standardization:")
print(dirty_data['country'].value_counts())
print("\nAfter standardization:")
print(dirty_data[['country_clean']].value_counts())

**Lựa chọn khác: Sử dụng khớp chuỗi mờ**

Đối với các trường hợp phức tạp hơn, chúng ta có thể sử dụng khớp chuỗi mờ với thư viện `rapidfuzz` để tự động phát hiện các chuỗi tương tự:


In [ ]:
try:
    from rapidfuzz import process, fuzz
except ImportError:
    print("rapidfuzz is not installed. Please install it with 'pip install rapidfuzz' to use fuzzy matching.")
    process = None
    fuzz = None

# Get unique countries
unique_countries = dirty_data['country'].unique()

# For each country, find similar matches
if process is not None and fuzz is not None:
    print("Finding similar country names (similarity > 70%):")
    for country in unique_countries:
        matches = process.extract(country, unique_countries, scorer=fuzz.ratio, limit=3)
        # Filter matches with similarity > 70 and not identical
        similar = [m for m in matches if m[1] > 70 and m[0] != country]
        if similar:
            print(f"\n'{country}' is similar to:")
            for match, score, _ in similar:
                print(f"  - '{match}' (similarity: {score}%)")
else:
    print("Skipping fuzzy matching because rapidfuzz is not available.")

### 2. Phát hiện Giá trị Số Bất Thường (Outliers)

Xem xét cột `age`, chúng ta có một số giá trị đáng ngờ như 199 và -5. Hãy sử dụng các phương pháp thống kê để phát hiện những giá trị bất thường này.


In [ ]:
# Display basic statistics
print("Age column statistics:")
print(dirty_data['age'].describe())

# Identify impossible values using domain knowledge
print("\nRows with impossible age values (< 0 or > 120):")
impossible_ages = dirty_data[(dirty_data['age'] < 0) | (dirty_data['age'] > 120)]
print(impossible_ages[['customer_id', 'name', 'age']])

#### Sử dụng phương pháp IQR (Khoảng tứ phân vị)

Phương pháp IQR là một kỹ thuật thống kê mạnh mẽ để phát hiện giá trị ngoại lai, ít nhạy cảm hơn với các giá trị cực đoan:


In [ ]:
# Calculate IQR for age (excluding impossible values)
valid_ages = dirty_data[(dirty_data['age'] >= 0) & (dirty_data['age'] <= 120)]['age']

Q1 = valid_ages.quantile(0.25)
Q3 = valid_ages.quantile(0.75)
IQR = Q3 - Q1

# Define outlier bounds
lower_bound = Q1 - 1.5 * IQR
upper_bound = Q3 + 1.5 * IQR

print(f"IQR-based outlier bounds for age: [{lower_bound:.2f}, {upper_bound:.2f}]")

# Identify outliers
age_outliers = dirty_data[(dirty_data['age'] < lower_bound) | (dirty_data['age'] > upper_bound)]
print(f"\nRows with age outliers:")
print(age_outliers[['customer_id', 'name', 'age']])

#### Sử dụng phương pháp Z-Score

Phương pháp Z-Score xác định các giá trị ngoại lai dựa trên độ lệch chuẩn so với giá trị trung bình:


In [ ]:
try:
    from scipy import stats
except ImportError:
    print("scipy is required for Z-score calculation. Please install it with 'pip install scipy' and rerun this cell.")
else:
    # Calculate Z-scores for age, handling NaN values
    age_nonan = dirty_data['age'].dropna()
    zscores = np.abs(stats.zscore(age_nonan))
    dirty_data['age_zscore'] = np.nan
    dirty_data.loc[age_nonan.index, 'age_zscore'] = zscores

    # Typically, Z-score > 3 indicates an outlier
    print("Rows with age Z-score > 3:")
    zscore_outliers = dirty_data[dirty_data['age_zscore'] > 3]
    print(zscore_outliers[['customer_id', 'name', 'age', 'age_zscore']])

    # Clean up the temporary column
    dirty_data = dirty_data.drop('age_zscore', axis=1)

#### Xử lý giá trị ngoại lai

Khi đã phát hiện, giá trị ngoại lai có thể được xử lý theo nhiều cách:
1. **Loại bỏ**: Xóa các hàng chứa giá trị ngoại lai (nếu đó là lỗi)
2. **Giới hạn**: Thay thế bằng các giá trị biên
3. **Thay thế bằng NaN**: Xem như dữ liệu thiếu và sử dụng các kỹ thuật bù đắp
4. **Giữ lại**: Nếu đó là các giá trị cực đoan hợp lệ


In [ ]:
# Create a cleaned version by replacing impossible ages with NaN
dirty_data['age_clean'] = dirty_data['age'].apply(
    lambda x: np.nan if (x < 0 or x > 120) else x
)

print("Age column before and after cleaning:")
print(dirty_data[['customer_id', 'name', 'age', 'age_clean']])

### 3. Phát hiện các hàng gần giống nhau

Lưu ý rằng tập dữ liệu của chúng ta có nhiều mục nhập cho "John Smith" với các giá trị hơi khác nhau. Hãy xác định các mục có khả năng trùng lặp dựa trên sự tương đồng về tên.


In [ ]:
# First, let's look at exact name matches (ignoring extra whitespace)
dirty_data['name_normalized'] = dirty_data['name'].str.strip().str.lower()

print("Checking for duplicate names:")
duplicate_names = dirty_data[dirty_data.duplicated(['name_normalized'], keep=False)]
print(duplicate_names.sort_values('name_normalized')[['customer_id', 'name', 'age', 'country']])

#### Tìm kiếm các bản sao gần giống bằng cách so khớp mờ

Để phát hiện các bản sao phức tạp hơn, chúng ta có thể sử dụng so khớp mờ để tìm các tên tương tự:


In [ ]:
try:
    from rapidfuzz import process, fuzz

    # Function to find potential duplicates
    def find_near_duplicates(df, column, threshold=90):
        """
        Find near-duplicate entries in a column using fuzzy matching.
        
        Parameters:
        - df: DataFrame
        - column: Column name to check for duplicates
        - threshold: Similarity threshold (0-100)
        
        Returns: List of potential duplicate groups
        """
        values = df[column].unique()
        duplicate_groups = []
        checked = set()
        
        for value in values:
            if value in checked:
                continue
                
            # Find similar values
            matches = process.extract(value, values, scorer=fuzz.ratio, limit=len(values))
            similar = [m[0] for m in matches if m[1] >= threshold]
            
            if len(similar) > 1:
                duplicate_groups.append(similar)
                checked.update(similar)
        
        return duplicate_groups

    # Find near-duplicate names
    duplicate_groups = find_near_duplicates(dirty_data, 'name', threshold=90)

    print("Potential duplicate groups:")
    for i, group in enumerate(duplicate_groups, 1):
        print(f"\nGroup {i}:")
        for name in group:
            matching_rows = dirty_data[dirty_data['name'] == name]
            print(f"  '{name}': {len(matching_rows)} occurrence(s)")
            for _, row in matching_rows.iterrows():
                print(f"    - Customer {row['customer_id']}: age={row['age']}, country={row['country']}")
except ImportError:
    print("rapidfuzz is not installed. Skipping fuzzy matching for near-duplicates.")

#### Xử lý các bản sao

Sau khi xác định, bạn cần quyết định cách xử lý các bản sao:
1. **Giữ lần xuất hiện đầu tiên**: Sử dụng `drop_duplicates(keep='first')`
2. **Giữ lần xuất hiện cuối cùng**: Sử dụng `drop_duplicates(keep='last')`
3. **Tổng hợp thông tin**: Kết hợp thông tin từ các hàng bị trùng lặp
4. **Xem xét thủ công**: Đánh dấu để con người kiểm tra


In [ ]:
# Example: Remove duplicates based on normalized name, keeping first occurrence
cleaned_data = dirty_data.drop_duplicates(subset=['name_normalized'], keep='first')

print(f"Original dataset: {len(dirty_data)} rows")
print(f"After removing name duplicates: {len(cleaned_data)} rows")
print(f"Removed: {len(dirty_data) - len(cleaned_data)} duplicate rows")

print("\nCleaned dataset:")
print(cleaned_data[['customer_id', 'name', 'age', 'country_clean']])

### Tóm tắt: Hoàn thiện Quy trình Làm sạch Dữ liệu

Hãy cùng tổng hợp tất cả thành một quy trình làm sạch dữ liệu toàn diện:


In [ ]:
def clean_dataset(df):
    """
    Comprehensive data cleaning function.
    """
    # Create a copy to avoid modifying the original
    cleaned = df.copy()
    
    # 1. Standardize categorical values (country)
    country_mapping = {
        'usa': 'USA', 'u.s.a': 'USA', 'united states': 'USA',
        'uk': 'UK', 'united kingdom': 'UK',
        'canada': 'Canada', 'mexico': 'Mexico'
    }
    cleaned['country'] = cleaned['country'].str.lower().map(country_mapping)
    
    # 2. Clean abnormal age values
    cleaned['age'] = cleaned['age'].apply(
        lambda x: np.nan if (x < 0 or x > 120) else x
    )
    
    # 3. Remove near-duplicate names (normalize whitespace)
    cleaned['name'] = cleaned['name'].str.strip()
    cleaned = cleaned.drop_duplicates(subset=['name'], keep='first')
    
    return cleaned

# Apply the cleaning pipeline
final_cleaned_data = clean_dataset(dirty_data)

print("Before cleaning:")
print(f"  Rows: {len(dirty_data)}")
print(f"  Unique countries: {dirty_data['country'].nunique()}")
print(f"  Invalid ages: {((dirty_data['age'] < 0) | (dirty_data['age'] > 120)).sum()}")

print("\nAfter cleaning:")
print(f"  Rows: {len(final_cleaned_data)}")
print(f"  Unique countries: {final_cleaned_data['country'].nunique()}")
print(f"  Invalid ages: {((final_cleaned_data['age'] < 0) | (final_cleaned_data['age'] > 120)).sum()}")

print("\nCleaned dataset:")
print(final_cleaned_data[['customer_id', 'name', 'age', 'country', 'purchase_amount']])

### 🎯 Bài Tập Thử Thách

Bây giờ đến lượt bạn! Dưới đây là một dòng dữ liệu mới với nhiều vấn đề về chất lượng. Bạn có thể:

1. Xác định tất cả các vấn đề trong dòng dữ liệu này
2. Viết mã để xử lý từng vấn đề
3. Thêm dòng dữ liệu đã được làm sạch vào tập dữ liệu

Dưới đây là dữ liệu có vấn đề:


In [ ]:
# New problematic row
new_row = pd.DataFrame({
    'customer_id': [13],
    'name': ['  Diana  Prince  '],  # Extra whitespace
    'age': [250],  # Impossible age
    'country': ['U.S.A.'],  # Inconsistent format
    'purchase_amount': [150.00]
})

print("New row to clean:")
print(new_row)

# TODO: Your code here to clean this row
# Hints:
# 1. Strip whitespace from the name
# 2. Check if the name is a duplicate (Diana Prince already exists)
# 3. Handle the impossible age value
# 4. Standardize the country name

# Example solution (uncomment and modify as needed):
# new_row_cleaned = new_row.copy()
# new_row_cleaned['name'] = new_row_cleaned['name'].str.strip()
# new_row_cleaned['age'] = np.nan  # Invalid age
# new_row_cleaned['country'] = 'USA'  # Standardized
# print("\nCleaned row:")
# print(new_row_cleaned)

### Những điểm chính

1. **Danh mục không đồng nhất** thường xuất hiện trong dữ liệu thực tế. Hãy luôn kiểm tra các giá trị duy nhất và chuẩn hóa chúng bằng cách sử dụng ánh xạ hoặc so khớp mờ.

2. **Giá trị ngoại lai** có thể ảnh hưởng lớn đến phân tích của bạn. Sử dụng kiến thức chuyên môn kết hợp với các phương pháp thống kê (IQR, Z-score) để phát hiện chúng.

3. **Các giá trị gần giống nhau** khó phát hiện hơn các giá trị trùng lặp hoàn toàn. Hãy cân nhắc sử dụng so khớp mờ và chuẩn hóa dữ liệu (chuyển chữ thường, loại bỏ khoảng trắng) để nhận diện chúng.

4. **Làm sạch dữ liệu là một quá trình lặp lại**. Bạn có thể cần áp dụng nhiều kỹ thuật và xem xét kết quả trước khi hoàn thiện tập dữ liệu đã làm sạch.

5. **Ghi lại các quyết định của bạn**. Theo dõi các bước làm sạch bạn đã áp dụng và lý do, vì điều này rất quan trọng để đảm bảo tính tái lập và minh bạch.

> **Thực hành tốt nhất:** Luôn giữ một bản sao của dữ liệu "bẩn" ban đầu. Không bao giờ ghi đè lên các tệp dữ liệu nguồn - hãy tạo các phiên bản đã làm sạch với cách đặt tên rõ ràng như `data_cleaned.csv`.



---

**Tuyên bố miễn trừ trách nhiệm**:  
Tài liệu này đã được dịch bằng dịch vụ dịch thuật AI [Co-op Translator](https://github.com/Azure/co-op-translator). Mặc dù chúng tôi cố gắng đảm bảo độ chính xác, xin lưu ý rằng các bản dịch tự động có thể chứa lỗi hoặc không chính xác. Tài liệu gốc bằng ngôn ngữ bản địa nên được coi là nguồn thông tin chính xác nhất. Đối với thông tin quan trọng, nên sử dụng dịch vụ dịch thuật chuyên nghiệp bởi con người. Chúng tôi không chịu trách nhiệm về bất kỳ sự hiểu lầm hoặc diễn giải sai nào phát sinh từ việc sử dụng bản dịch này.
